cpp/src/io/orc/writer_impl.cu

```diff
@@ -297,51 +335,56 @@ void writer::impl::init_dictionaries(orc_column_view *columns,
+                                      host_span<stripe_rowgroups const> stripe_bounds,
                                       hostdevice_vector<gpu::DictionaryChunk> const &dict,
                                       uint32_t *dict_index,
                                       hostdevice_vector<gpu::StripeDictionary> &stripe_dict)
 {
   const auto num_rowgroups = dict.size() / str_col_ids.size();
 
-  for (size_t i = 0; i < str_col_ids.size(); i++) {
-    size_t direct_cost = 0, dict_cost = 0;
-    auto &str_column = columns[str_col_ids[i]];
+  for (size_t col_idx = 0; col_idx < str_col_ids.size(); ++col_idx) {
+    size_t direct_cost = 0;
+    size_t dict_cost   = 0;
+    auto &str_column   = columns[str_col_ids[col_idx]];
     str_column.attach_stripe_dict(stripe_dict.host_ptr(), stripe_dict.device_ptr());
 
-    for (size_t j = 0, g = 0; j < stripe_list.size(); j++) {
-      const auto num_chunks = stripe_list[j];
-      auto *sd              = &stripe_dict[j * str_col_ids.size() + i];
-      sd->column_data_base  = str_column.host_dict_chunk(0)->column_data_base;
-      sd->dict_data         = str_column.host_dict_chunk(g)->dict_data;
-      sd->dict_index        = dict_index + i * num_rows;  // Indexed by abs row
-      sd->column_id         = str_col_ids[i];
-      sd->start_chunk       = (uint32_t)g;
-      sd->num_chunks        = num_chunks;
-      sd->num_strings       = 0;
-      sd->dict_char_count   = 0;
-      for (size_t k = g; k < g + num_chunks; k++) {
-        const auto &dt = dict[k * str_col_ids.size() + i];
-        sd->num_strings += dt.num_dict_strings;
-        direct_cost += dt.string_char_count;
-        dict_cost += dt.dict_char_count + dt.num_dict_strings;
-      }
-
-      g += num_chunks;
+    for (auto const &stripe : stripe_bounds) {
+      auto &sd            = stripe_dict[stripe.id * str_col_ids.size() + col_idx];
+      sd.column_data_base = str_column.host_dict_chunk(0)->column_data_base;
+      sd.dict_data        = str_column.host_dict_chunk(stripe.first)->dict_data;
+      sd.dict_index       = dict_index + col_idx * str_column.data_count();  // Indexed by abs row
+      sd.column_id        = str_col_ids[col_idx];
+      sd.start_chunk      = stripe.first;
+      sd.num_chunks       = stripe.size;
+      sd.dict_char_count  = 0;
+      sd.num_strings =
+        std::accumulate(stripe.cbegin(), stripe.cend(), 0, [&](auto dt_str_cnt, auto rg_idx) {
+          const auto &dt = dict[rg_idx * str_col_ids.size() + col_idx];
+          return dt_str_cnt + dt.num_dict_strings;
+        });
+      direct_cost = std::accumulate(
+        stripe.cbegin(), stripe.cend(), direct_cost, [&](auto dir_cost, auto rg_idx) {
+          const auto &dt = dict[rg_idx * str_col_ids.size() + col_idx];
+          return dir_cost + dt.string_char_count;
+        });
+      dict_cost =
+        std::accumulate(stripe.cbegin(), stripe.cend(), dict_cost, [&](auto dt_cost, auto rg_idx) {
```

Comment:
Moved the cost accumulation out of the loop, also now it's only done when dictionary encoding is enabled.
The code would be a bit cleaner if we used 2dspan for dict (+transpose to avoid the stride), but I'd like to leave this for another PR. Let me know what you thing about the current code.